book/visual-effects.md

```diff
@@ -5,312 +5,396 @@ prev: security
-        skia_bytes = skia_image.tobytes()
-        rect = sdl2.SDL_Rect(0, 0, WIDTH, HEIGHT)
-        skia_surface = Browser.to_sdl_surface(skia_bytes)
-        sdl2.SDL_BlitSurface(
-            skia_surface, rect, self.window_surface, rect)
-        sdl2.SDL_UpdateWindowSurface(self.sdl_window)
+class Browser:
+    def draw(self):
+        # ...
+        draw_rect(self.skia_surface, 10, 50, 35, 90)
+        path = skia.Path().moveTo(15, 70).lineTo(30, 55).lineTo(30, 85)
+        paint = skia.Paint(Color=skia.ColorBlack, Style=skia.Paint.kFill_Style)
+        with surface as canvas:
+            canvas.drawPath(path, paint)
 ```
 
-And here is the `to_sdl_surface` method:
+Once this is done, we need to copy from the Skia surface to the SDL window:
 
 ``` {.python}
-    def to_sdl_surface(skia_bytes):
-        depth = 32 # 4 bytes per pixel
-        pitch = 4 * WIDTH # 4 * WIDTH pixels per line on-screen
-        # Skia uses an ARGB format - alpha first byte, then
-        # through to blue as the last byte.
-        alpha_mask = 0xff000000
-        red_mask = 0x00ff0000
-        green_mask = 0x0000ff00
-        blue_mask = 0x000000ff
-        return sdl2.SDL_CreateRGBSurfaceFrom(
-            skia_bytes, WIDTH, HEIGHT, depth, pitch,
-            red_mask, green_mask, blue_mask, alpha_mask)
+class Browser:
+    def draw(self):
+        # ...
+        self.commit_canvas()
 ```
 
-Finally, `handle_enter` and `handle_down` no longer need an event parameter.
-
-[^surface]: In Skia and SDL, a *surface* is a representation of a graphics buffer
-into which you can draw "pixels" (bits representing colors). A surface may or
-may not be bound to the actual pixels on the screen via a window, and there can
-be many surfaces. A *canvas* is an API interface that allows you to draw
-into a surface with higher-level commands such as for rectangles or text. In
-our implementation, we'll start with separate surfaces for Skia and SDL for
-simplicity. In a highly optimized browser, minimizing the number of surfaces
-is important for good performance.
-
-In the `Tab` class, the differences in `draw` is the new `surface` parameter
-that gets passed around, and using the Skia API to measure font metrics. Skia's
-`measureText` method on a font is the same as the `measure` method on a Tkinter
-font.
+We've only got a few minor changes left elsewhere in the browser.
+`Tab` also has a `draw` method and draws a cursor; it needs to use
+`draw_line` for that:
 
 ``` {.python}
+class Tab:
     def draw(self, surface):
-        # ...
+        is self.focus:
+            # ...
             x = obj.x + obj.font.measureText(text)
+            y = obj.y - self.scroll + CHROME_PX
+            draw_line(surface, x, y, x, y + obj.height)
 ```
 
-Update all the other places that `measure` was called to use the Skia method
-(and also create Skia fonts instead of Tkinter ones, of course).
+Note that I've renamed `draw`'s argument from `surface` to `canvas`
```

Comment:
You didn't do the rename....

In any case, to answer question 1, yes you can pass the canvas around. I think it's not a good idea
to keep it bound in a variable, but keeping it around for the duration of Browser.draw is a good
simplification.

Reply:
Ok, will do